akka-stream/src/main/scala/akka/stream/impl/FoldResourceSinkAsync.scala

```diff
@@ -0,0 +1,162 @@
+      }.invoke _
+
+      override def preStart(): Unit = createResource()
+
+      override def onPush(): Unit =
+        blockingStream match {
+          case Some(resource) ⇒
+            try {
+              write(resource, grab(in)).onComplete(pushCallback)(sameThreadExecutionContext)
+            } catch errorHandler
+          case None ⇒
+          // we got a pull but there is no open resource, we are either
+          // currently creating/restarting then the pull will be triggered when creating the
+          // resource completes, or shutting down and then the push does not matter anyway
+        }
+
+      override def postStop(): Unit = {
+        val o = blockingStream
+        blockingStream = None
+        o match {
+          case Some(s) ⇒
+            writeFail.foreach(failStage)
+
+            try {
+              val f = close(s)
+              writeFail match {
+                case Some(ex) ⇒ promise.tryCompleteWith(f.flatMap { _ ⇒ Future.failed(ex) })
+                case None ⇒ promise.tryCompleteWith(f.map(_ ⇒ Done).recoverWith {
+                  case NonFatal(ex) ⇒
+                    failStage(ex)
+                    Future.failed(ex)
+                })
+              }
+            } catch {
+              case NonFatal(ex) ⇒
+                if (writeFail.isEmpty) failStage(ex)
+                promise.tryFailure(writeFail.getOrElse(ex))
+            }
+
+          case None ⇒ writeFail match {
+            case None     ⇒ promise.trySuccess(Done)
+            case Some(ex) ⇒ promise.tryFailure(ex)
+          }
+        }
+      }
+
+      private def restartResource(): Unit = {
+        val o = blockingStream
+        blockingStream = None
+
+        o match {
+          case Some(resource) ⇒
+            // wait for the resource to close before restarting
+            try {
+              close(resource).onComplete(getAsyncCallback[Try[Unit]] {
+                case Success(_) ⇒
+                  createResource()
+                case Failure(ex) ⇒
+                  promise.tryFailure(ex)
+                  failStage(ex)
+
+              }.invoke)
+            } catch {
+              case NonFatal(ex) ⇒
+                promise.tryFailure(ex)
+                failStage(ex)
+            }
+
+          case None ⇒ createResource()
+        }
+      }
+
+      private def createResource(): Unit = {
+        try {
+          create().onComplete { resource ⇒
+            createdCallback(resource).recover {
+              case _: StreamDetachedException ⇒
+                // stream stopped
+                resource match {
+                  case Success(r)  ⇒ close(r)
+                  case Failure(ex) ⇒ throw ex // failed to open but stream is stopped already
+                }
+            }
+          }
+        } catch {
+          case NonFatal(ex) ⇒
+            failStage(ex)
+            promise.tryFailure(ex)
+        }
+      }
+
+      setHandler(in, this)
+    }
```

Comment:
Missing a `postStop`, on an abrupt termination (shutting down the materializer while the stream is running) this will leak the resource.

Reply:
I have this: `       override def postStop(): Unit` non line 79.  Is there another place I need to put it or were you saying I missed something within my definition of postStop?  The test failure indicates you're right but I don't see what I'm doing wrong.

Reply:
Oh, sorry, my bad, I missed that!